### ISSUE TITLE
Port twisted.web.vhost to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8132]
|<img alt="jrolli's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| jrolli reported|
|-|-|
|Trac ID|trac#8132|
|Type|enhancement|
|Created|2015-11-29 06:58:07Z|
|Branch|https://github.com/twisted/twisted/tree/twvhost-py3-8132-3|

Relies on twisted.python.roots being ported to Python 3 ([#8131](https://github.com/twisted/twisted/issues/8131)).

Because twisted.web.vhost deals directly with the raw headers and URLs as transmitted via HTTP, it needs to explicitly use bytes for Python 3 when doing comparisons and string functions.

twisted.web.test.test_vhost needs to be updated after twisted.web.vhost to ensure bytes are compared to bytes and strings are compared to strings.

Attachments:

* [twisted_web_vhost_py3-8132.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d2c/d2ccec7748447440624b77f1d17d74a1c0e44779/84806cf6dad8de1a4953d4a77a2ae60bafb41854.patch) (5947 bytes) - added by jrolli on 2015-11-29 07:10:05Z - 
* [twisted_web_vhost_py3-8132-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d2c/d2ccec7748447440624b77f1d17d74a1c0e44779/bf3e7e527671cd6e45c2922b6cc3e90b3b6297bb.patch) (5785 bytes) - added by jrolli on 2015-12-06 07:10:42Z - 
* [twisted_web_vhost_py3-8132-rebased.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d2c/d2ccec7748447440624b77f1d17d74a1c0e44779/bb90e5334fb92847ddb786ad10d5fd5d79c78648.patch) (10906 bytes) - added by jrolli on 2016-01-11 16:37:18Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__8132 8132
type__enhancement enhancement
reporter__jrolli jrolli
priority__normal normal
milestone__Python_3_x Python-3.x
branch__branches_twvhost_py3_8132_3 branches/twvhost-py3-8132-3
branch_author__hawkowl__jrolli hawkowl, jrolli
status__closed closed
resolution__fixed fixed
component__web web
keywords__None None
time__1448780287751710 1448780287751710
changetime__1452592320231917 1452592320231917
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_3"></a>|@hawkowl commented|
|-|-|

(In [46379]) Branching to twvhost-py3-8132.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_4"></a>|@hawkowl commented|
|-|-|

(In [46380]) Apply patch from jrolli, refs [#8132](https://github.com/twisted/twisted/issues/8132)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl set owner to jrolli|
|-|-|

Hi jrolli,

Thanks for this.

There's a lot of missing coverage in this file (see https://codecov.io/github/twisted/twisted/twisted/web/vhost.py?ref=05470d6d0ec3e592cca940735d85c79a4d3e0d20 ) so some tests for the changed lines, ideally bringing it to 100%, would be excellent and make sure that the code changes are actually compatible on 3.

I've done some more changes as per [#8131](https://github.com/twisted/twisted/issues/8131), adding ```__future__``` lines, etc. As such, please base further patches off this branch, and note that it's off the branch when you put it up for review again.

Thanks again!

**Comment by twisted-trac** [comment]
|<img alt="jrolli's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|jrolli commented|
|-|-|

Second patch is based off of branch ```twvhost-py3-8132```.  Completes test coverage of twisted.web.vhost and adds ```setHost``` function to ```DummyRequest``` class in order to allow testing of ```VHostMonsterResource```.

**Comment by twisted-trac** [comment]
|<img alt="jrolli's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|jrolli removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_8"></a>|@hawkowl commented|
|-|-|

(In [46442]) apply patch from jrolli, refs [#8132](https://github.com/twisted/twisted/issues/8132)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_9"></a>|@hawkowl commented|
|-|-|

(In [46444]) Branching to twvhost-py3-8132-2.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_10"></a>|@hawkowl set owner to @hawkowl|
|-|-|

Hi jrolli, thanks for the patch!

Unfortunately ```DummyRequest``` still has a ```header``` attribute, which has been deprecated and now removed -- so it doesn't match the true Request. I have filed [#8140](https://github.com/twisted/twisted/issues/8140) and will be uploading a patch for that ASAP, so this is kind of blocked on it, as ```DummyRequest.getHeader```, which your patch uses, will not work as it looks at the wrong list of headers. This is the correct API for it to be using, but if the rest of your patch used the correct API (requestHeaders), ```getHeader``` will not look at it. Legacy :(

I will fix [#8140](https://github.com/twisted/twisted/issues/8140), amend this branch to use the correct API, and then get Adi, the other Twisted fellow to do a final review, as I won't be able to review it further under the review guidelines.

Thanks again for your work!

**Comment by twisted-trac** [comment]
|<img alt="jrolli's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|jrolli commented|
|-|-|

**Apply patch to ```trunk```, not existing branch**

Rebased patch to ```trunk``` because of significant conflicts with API updates to ```DummyRequest``` from [#8140](https://github.com/twisted/twisted/issues/8140).

Updated references to ```DummyRequest.headers``` in previous patches to utilize ```DummyRequest.requestHeaders.addRawHeader```.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_12"></a>|@hawkowl commented|
|-|-|

(In [46615]) Branching to twvhost-py3-8132-3.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_13"></a>|@hawkowl set status to `closed`|
|-|-|

(In [46618]) Merge twvhost-py3-8132-3: Port twisted.web.vhost to Python 3

Author: jrolli
Reviewers: hawkowl
Fixes: [#8132](https://github.com/twisted/twisted/issues/8132)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_14"></a>|@hawkowl commented|
|-|-|

Thanks, jrolli.
